React/NextJS routing.md

```diff
@@ -0,0 +1,428 @@
+In React you need to add a routing library to handle the routing, and you need to add the route components and handle the routes
+for the new pages manually.
+
+In NextJS the routing is handled in a simpler way, and there are few built-in tools to help you with that.
+
+The explanation for the routing and everything it goes with it is based on the official NextJS routing documentation. 
+For more information, see the [official NextJS routing documentation](https://nextjs.org/docs/routing/introduction).
+
+
+## The `pages` folder
+
+The `pages` folder contains all your pages (basically the files with the extracted React components) where each page is 
+also a route.
+
+E.g. the `hello-world.tsx` file in the root of the `pages` folder will be a `your.domain/hello-world` route.
+More examples:
+```javascript
+pages/about.tsx => your.domain/about
+pages/ABOUT.tsx => your.domain/ABOUT
+
+pages/company/index.tsx => your.comain/company
+pages/company/info.tsx => your.domain/company/info
+pages/company/contact-us.tsx => your.domain/company/contact-us
+```
+
+These are also called the predefined pages/routes.
+Note how the first two examples are case-sensitive.
+
+### Dynamic routes
+
+Usually we also want to achieve something like this for our pages:
+```
+your.domain/article/New-React-18
+your.domain/article/How-To-Handle-Routing-In-NextJS
+```
+and luckily NextJS provides an easy way to support the dynamic routes.
+
+To create a page that is dynamic you need to wrap the brackets around the page name:
+```
+pages/article/[articleId].tsx
+// or pages/article/[article].tsx
+// or pages/article/[id].tsx
+```
+Both of the above examples will be a match for this page, so the code for that page will be executed. On the page we can
+then extract the path by using a built-in hook `useRouter` (more details about the hook later).
+
+```jsx
+// pages/article/[articleId].tsx
+
+import { useRouter } from 'next/router';
+
+const ArticlePage = () => {
+    const { query } = useRouter()
+
+    if (!query.articleId) {
+        return <div>loading...</div>
+    }
+
+    return <div>{query.articleId}</div>;
+}
+
+export default ArticlePage;
+```
+
+The example above shows a statically optimized page made by
+[Automatic Static Optimization (ASG)](https://nextjs.org/docs/advanced-features/automatic-static-optimization),
+so to get a path you need to use `useRouter`.
+[Server Side Rendering (SSR)](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props) and 
+[Static Site Generation (SSG)](https://nextjs.org/docs/basic-features/data-fetching/get-static-props) can also give you the information from `useRouter`,
+but they also propagate that info (and more) via context.
+
+```tsx
+// SSR pages/article/[articleId].tsx
+
+import { useRouter } from 'next/router';
+
+const ArticlePage = ({ articleId }) => {
+    const { query } = useRouter()
+
+    if (!query.articleId) {
```

Comment:
You don't need this if because you are getting `articleId` form the props, so you don't need `useRouter` in this example

Reply:
it's not a real world example, also it's a bad practice, so I'll remove the code

Reply:
Will do.